sw/device/tests/pwm_smoketest.c

```diff
@@ -0,0 +1,104 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_gpio.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_pwm.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/test_framework/ottf_utils.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+static volatile const uint8_t kClocksHz[] = {20, 50, 100, 200};
+static volatile const uint8_t kDutyCycles[] = {10, 30, 50, 70, 90};
+
+static dif_pwm_config_t compute_clk_config(uint32_t pwm_clk) {
+  enum {
+    // Theres a trade off on the precision of the clock and the precision of the
+    // dutycycle, the higher the kDutyCycleResulution the more precise is the
+    // the dutycycle and less precise is the pwm frequency.
+    // By experimentation kDutyCycleResulution would generate an error of 2.5%
+    // on the frequency and 6% on the dutycycle.
+    kDutyCycleResulution = 4,
+    kBeatsPerCycle = 2 << (kDutyCycleResulution + 1),  // 2 ^ (DC_RESN + 1)
+  };
+  return (dif_pwm_config_t){
+      .beats_per_pulse_cycle = kBeatsPerCycle,
+      .clock_divisor =
+          ((uint32_t)kClockFreqAonHz / (kBeatsPerCycle * pwm_clk)) - 1};
+}
+
+bool test_main(void) {
+  dif_pwm_t pwm;
+  mmio_region_t addr = mmio_region_from_addr(TOP_EARLGREY_PWM_AON_BASE_ADDR);
+  CHECK_DIF_OK(dif_pwm_init(addr, &pwm));
+
+  dif_pinmux_t pinmux;
+  addr = mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR);
+  CHECK_DIF_OK(dif_pinmux_init(addr, &pinmux));
+  CHECK_DIF_OK(dif_pinmux_output_select(&pinmux, kTopEarlgreyPinmuxMioOutIoa8,
+                                        kTopEarlgreyPinmuxOutselPwmAonPwm0));
+
+  dif_gpio_t gpio;
+  addr = mmio_region_from_addr(TOP_EARLGREY_GPIO_BASE_ADDR);
+  CHECK_DIF_OK(dif_gpio_init(addr, &gpio));
+  CHECK_DIF_OK(dif_gpio_output_set_enabled_all(&gpio, 0x1));
+
+  LOG_INFO("Use IOA7 to let host know when sleep is active.");
```

Comment:
I think this message maybe needs altering since this test doesn't enter sleep unlike the chip-level DV one. Also, if I'm understanding correctly, the host is using that GPIO signal to indicate to the device test code on the Ibex that it is sampling?